### ISSUE TITLE
How many epoch should be great for Training on ImageNet datasets?

### ISSUE THREAD

**Issue by NovaMind-Z** [google-research/fixmatch#21]
Hi, 
After implementing your brilliant work to PyTorch and testing it on Cifar10, recently when I try to reproduce the results on ImageNet. And I am puzzled with the setting of epochs. From your paper, I find that the number of epoch is 300 for unsupervised data, and we know that the amount of the unsupervised data is 10 times as large as that of supervised data, and in each batch the amount of unsupervised data is 5 times(ulratio=5) as large as supervised's, so I guess the number of epoch for supervised data should be 300*10/5 = 600？ However, in the Readme.txt of your code, I find that you set 3000 epochs for supervised data, why? BTW, I am wondering how long the training process will last? I guess 3000 epochs will cost really long time.  

Thanks in advance for your reply

**Comment by AlexeyKurakin** [comment]
Hi,

Sorry if it caused confusion.
In the code number of training epochs is measured in epochs of supervised data. So for 10% imagenet experiments it's 3000 epochs.
What I actually wanted to say in the paper is that if we count number of examples then 3000 epochs on 10% supervised imagenet will be equivalent [in terms of number of examples] to 300 epochs of unsupervised data [because unsupervised data is 100% imagenet].

**Comment by NovaMind-Z** [comment]
> Hi,
> 
> Sorry if it caused confusion.
> In the code number of training epochs is measured in epochs of supervised data. So for 10% imagenet experiments it's 3000 epochs.
> What I actually wanted to say in the paper is that if we count number of examples then 3000 epochs on 10% supervised imagenet will be equivalent [in terms of number of examples] to 300 epochs of unsupervised data [because unsupervised data is 100% imagenet].

I do appreciate your reply! Could you please share me how many hours should it be for the whole training process? I tried 300 epochs training using 32 GPUs and placed 8 supervised images on each of them (I guess the equivalent supervised batch size is 32*8 = 256), it cost me around 50 hours and only reach the 64% for the top1 acc. So 3000 epochs should cost 500 hours? I wonder how long you have spent for the training so that I can check whether my code is slow.

**Comment by AlexeyKurakin** [comment]
We trained on TPU with 32 cores, which should be roughly equivalent in terms of compute to 32 v100 [but in practice there might be some other differences in performance]. However we used larger batch size - 32 supervised images per core. In our setup it took about 2.5 - 3 days to train for full 3000 epochs.

If you want to accelerate training, consider adjust learning rate schedule and train for smaller number of epochs. For example you can train for 1500 epochs and decay learning rate every 300 epochs instead of 600 epochs. I think you final accuracy will be slightly lower in such case, however you will accelerate training.
Also we noticed that ImageNet performance somewhat sensitive to learning rate schedule, so you actually might same (or even better) as our result with less number of epochs if you willing to spend time on tweaking learning rate schedule.

Also suggestion regarding batch size. Batch norm may not perform well if your batch size is too small. In particular 8 per GPU might be too small and you will benefit by increasing it to either 16 or 32 (depending on what fits into GPU memory). Alternatively you can do global batch norm (which aggregates batches from all GPUs when computing batch norm statistics).

**Comment by NovaMind-Z** [comment]
> We trained on TPU with 32 cores, which should be roughly equivalent in terms of compute to 32 v100 [but in practice there might be some other differences in performance]. However we used larger batch size - 32 supervised images per core. In our setup it took about 2.5 - 3 days to train for full 3000 epochs.
> 
> If you want to accelerate training, consider adjust learning rate schedule and train for smaller number of epochs. For example you can train for 1500 epochs and decay learning rate every 300 epochs instead of 600 epochs. I think you final accuracy will be slightly lower in such case, however you will accelerate training.
> Also we noticed that ImageNet performance somewhat sensitive to learning rate schedule, so you actually might same (or even better) as our result with less number of epochs if you willing to spend time on tweaking learning rate schedule.
> 
> Also suggestion regarding batch size. Batch norm may not perform well if your batch size is too small. In particular 8 per GPU might be too small and you will benefit by increasing it to either 16 or 32 (depending on what fits into GPU memory). Alternatively you can do global batch norm (which aggregates batches from all GPUs when computing batch norm statistics).

Thanks! The information you provided is really helpful. I think I can slim my code and take a fast try.  
Once again, thanks for both your brilliant job and your patient explaination.
